rust/src/mysql/parser.rs

```diff
@@ -0,0 +1,3447 @@
+    let payload =
+        unsafe { std::slice::from_raw_parts(header.payload.as_ptr(), header.payload.len()) };
+    let (i, response_code) = be_u8(payload)?;
+    match response_code {
+        // OK
+        0x00 => {
+            let (_, response) = parse_response_ok(i)?;
+            Ok((rem, response))
+        }
+        // ERR
+        0xFF => {
+            let (_, response) = parse_response_err(i)?;
+            Ok((rem, response))
+        }
+        _ => parse_response_binary_resultset(rem, response_code as u64),
+    }
+}
+
+pub fn parse_stmt_fetch_response(i: &[u8]) -> IResult<&[u8], MysqlResponse> {
+    let (rem, header) = parse_packet_header(i)?;
+    let payload =
+        unsafe { std::slice::from_raw_parts(header.payload.as_ptr(), header.payload.len()) };
+    let (i, response_code) = be_u8(payload)?;
+    match response_code {
+        // OK
+        0x00 => {
+            let (_, response) = parse_response_ok(i)?;
+            Ok((rem, response))
+        }
+        // ERR
+        0xFF => {
+            let (_, response) = parse_response_err(i)?;
+            Ok((rem, response))
+        }
+        _ => parse_response_binary_resultset(rem, response_code as u64),
+    }
+}
+
+pub fn parse_auth_request(i: &[u8]) -> IResult<&[u8], ()> {
+    let (rem, _header) = parse_packet_header(i)?;
+    Ok((rem, ()))
+}
+
+pub fn parse_auth_responsev2(i: &[u8]) -> IResult<&[u8], MysqlResponse> {
+    let (rem, header) = parse_packet_header(i)?;
+    let payload =
+        unsafe { std::slice::from_raw_parts(header.payload.as_ptr(), header.payload.len()) };
+    let (i, response_code) = be_u8(payload)?;
+    match response_code {
+        // OK
+        0x00 => {
+            let (_, response) = parse_response_ok(i)?;
+            Ok((rem, response))
+        }
+        // auth data
+        _ => Ok((
+            rem,
+            MysqlResponse {
+                item: MysqlResponsePacket::AuthData,
+            },
+        )),
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use super::*;
+
+    #[test]
+    fn test_parse_packet_header() {
+        let pkt: &[u8] = &[0x07, 0x00, 0x00, 0x03];
+        let (rem, packet_header) = parse_packet_header(pkt).unwrap();
+
+        assert!(rem.is_empty());
+        assert_eq!(packet_header.pkt_len, 7);
+        assert_eq!(packet_header.pkt_num, 3);
+    }
+
+    #[test]
+    fn test_parse_handshake_request() {
+        let pkt: &[u8] = &[
+            0x49, 0x00, 0x00, 0x00, 0x0a, 0x38, 0x2e, 0x34, 0x2e, 0x30, 0x00, 0x51, 0x00, 0x00,
+            0x00, 0x3e, 0x7d, 0x6a, 0x6a, 0x1a, 0x2d, 0x2b, 0x6b, 0x00, 0xff, 0xff, 0xff, 0x02,
+            0x00, 0xff, 0xdf, 0x15, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0x61, 0x74, 0x54, 0x07, 0x62, 0x28, 0x5d, 0x21, 0x06, 0x44, 0x06, 0x62, 0x00, 0x63,
+            0x61, 0x63, 0x68, 0x69, 0x6e, 0x67, 0x5f, 0x73, 0x68, 0x61, 0x32, 0x5f, 0x70, 0x61,
+            0x73, 0x73, 0x77, 0x6f, 0x72, 0x64, 0x00,
```

Comment:
As a good practice, when we have buffers of hex, especially long ones like that, could you add a comment indicating what is the buffer content?